### PR TITLE
fix(service-worker): allow service worker to handle non root base hrefs

### LINKED ISSUE #25055: Service Worker Http Error 504 When Build Using --base-href=./ (Dynamic Base Href)
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

Whenever the application load for first time, and then the network goes offline, the service worker couldn't provide the application cache and failed with http error 504 as shown below:
service-worker-504

## Expected behavior

The service worker could provide the application cache and the application could still be loaded even when the network goes offline.

## Minimal reproduction of the problem with instructions

1. `ng new sw2 --routing`
2.  `cd sw2`
3. `ng add @angular/pwa --project sw2`
4. `ng build --prod --base-href=./`
5. `cd dist`
6. `cd sw2`
7. `npm install -g http-server` (you can use any server to test this, if you prefer another one)
8. `http-server`
9. Access `localhost:8080` (8080 is the default port that http-server use at the time of reporting this issue)
10. Stop the server
11. Now try reloading the app again, and the browser will display the message:
```
This page isn’t working
localhost took too long to respond.
HTTP ERROR 504
```

## What is the motivation / use case for changing the behavior?

I need to build my angular application with dynamic base href, because my application will be deployed as any name context e. g. client1app, client2app, /external/client3app, etc. 
The application should be working good even when the network is offline (still loaded correctly with cached resources and data).

## Environment
<pre><code>
Angular version: 6.0.9
Angular CLI version: 6.0.8

Browser:
- [x] Chrome (desktop) version 67.0.3396.99 (Official Build) (64-bit)
 
For Tooling issues:
- Node version: v8.9.4 
- Platform: Windows 7 
</code></pre>

### LINKED ISSUE #30505: Angular service worker state (/ngsw/state) doesn't work if context is not /
# 🐞 bug report

### Affected Package
@angular/pwa
@angular/service-worker

### Description
Service worker state is not viewable if the app context is not /

## 🔬 Minimal Reproduction
* build the angular PWA with `--base-href="/myfolder/"`
* deploy the app to the myfolder subfolder
* check that https://server/myfolder/ngsw/state does not work

For example:
* go to [https://www.bassanofamiglia.it/ng-garsiawe/portalecittadino](url) to see the app
* check that [https://www.bassanofamiglia.it/ng-garsiawe/portalecittadino/ngsw/state](url) doesn't work

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 7.3.8
Node: 10.15.0
OS: win32 x64
Angular: 7.2.12
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.12.4
@angular-devkit/build-angular      0.12.4
@angular-devkit/build-ng-packagr   0.12.4
@angular-devkit/build-optimizer    0.12.4
@angular-devkit/build-webpack      0.12.4
@angular-devkit/core               7.2.4
@angular-devkit/schematics         7.3.8
@angular/cdk                       7.3.7
@angular/cli                       7.3.8
@angular/material                  7.3.7
@angular/pwa                       0.12.4
@angular/service-worker            7.2.15
@ngtools/json-schema               1.1.0
@ngtools/webpack                   7.2.4
@schematics/angular                7.3.8
@schematics/update                 0.13.8
ng-packagr                         4.7.1
rxjs                               6.4.0
typescript                         3.2.4
webpack                            4.28.4

</code></pre>

this comment to another issue is relevant: https://github.com/angular/angular/issues/30351#issuecomment-492754624

### PR DESCRIPTION
**Pull request by sonukapoor**
This PR is based on the work from @mattlewis92. For now, I just have rebased and fixed the conflicts. I have not addressed any comments from the previous PR yet. 

See https://github.com/angular/angular/pull/31082 for more details on his PR.

Closes #25055
Closes #30505

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by mattlewis92** [comment]
Awesome, thanks for taking this over @sonukapoor 🙌

File: packages/service-worker/worker/test/happy_spec.ts [hunk 1]
```diff
@@ -1515,4 +1592,4 @@ function makeNavigationRequest(
     mode: 'navigate',
     ...init,
   });
-}
+}
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts:1595]: 🙅

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx for taking this over, @sonukapoor!
I did a preliminary review and left some comments. Can you also split this up into two commits (as described in https://github.com/angular/angular/pull/31082#pullrequestreview-256034185).

**Comment by gkalpak** [comment]
Thx, @sonukapoor, for moving this further along. As discussed "offline", I am taking this over.

This is now superceded by #37922.

**Pull request closed without merging**